prov/sm2/src/sm2_progress.c

```diff
@@ -42,6 +42,193 @@
+		host_to_dev_data->rx_flags = rx_entry->flags;
+
+		ipc_info = &host_to_dev_data->ipc_info;
+		ipc_info->iface = iface;
+		ipc_info->device = device;
+
+		/* Open IPC handle */
+		ret = ofi_hmem_get_base_addr(ipc_info->iface, device_ptr, &base,
+					     &ipc_info->base_length);
+		if (ret) {
+			FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+				"Failed to get device memory base address. "
+				"Error code: %d\n",
+				ret);
+			return ret;
+		}
+
+		ret = ofi_hmem_get_handle(ipc_info->iface, base,
+					  ipc_info->base_length,
+					  (void **) &ipc_info->ipc_handle);
+		if (ret) {
+			FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+				"Failed to open IPC handle in host to device "
+				"protocol. Error code: %d\n",
+				ret);
+			return ret;
+		}
+
+		ipc_info->base_addr = (uintptr_t) base;
+		ipc_info->offset = (uintptr_t) device_ptr -
+				   (uintptr_t) ipc_info->base_addr;
+
+		/* Send xfer_entry with IPC handle back to sender */
+		sm2_fifo_write(ep, sender_gid, xfer_entry);
+
+		return FI_SUCCESS;
+	} else {
+		assert(1 == 0);
+		FI_WARN(&sm2_prov, FI_LOG_EP_CTRL, "CMA error %d\n", errno);
+		return -FI_EIO;
+	}
+}
+
+static int sm2_ipc_host_to_dev_hmem_memcpy(struct sm2_ep *ep,
+					   struct sm2_xfer_entry *xfer_entry)
+{
+	struct sm2_domain *domain;
+	struct sm2_ipc_host_to_dev_data *host_to_dev_data;
+	struct ipc_info *ipc_info;
+	struct ofi_mr_entry *mr_entry;
+	void *dest;
+	int ret;
+
+	host_to_dev_data =
+		(struct sm2_ipc_host_to_dev_data *) xfer_entry->user_data;
+	ipc_info = &host_to_dev_data->ipc_info;
+
+	domain = container_of(ep->util_ep.domain, struct sm2_domain,
+			      util_domain);
+
+	ret = ofi_ipc_cache_search(domain->ipc_cache,
+				   xfer_entry->hdr.sender_gid, ipc_info,
+				   &mr_entry);
+	if (ret)
+		return ret;
+
+	dest = (char *) (uintptr_t) mr_entry->info.ipc_mapped_addr +
+	       (uintptr_t) ipc_info->offset;
+
+	ret = ofi_copy_to_hmem(ipc_info->iface, ipc_info->device, dest,
+			       host_to_dev_data->host_ptr,
+			       xfer_entry->hdr.size);
+
+	ofi_mr_cache_delete(domain->ipc_cache, mr_entry);
+
+	if (ret != 0)
+		return ret;
+
+	return FI_SUCCESS;
+}
+
+static inline int sm2_cma_loop(struct sm2_ep *ep,
+			       struct sm2_xfer_entry *xfer_entry,
+			       struct fi_peer_rx_entry *rx_entry, size_t total,
+			       bool write, bool *ipc_host_to_dev)
+{
+	int err;
+	ssize_t ret;
+	struct sm2_av *sm2_av =
+		container_of(ep->util_ep.av, struct sm2_av, util_av);
+	struct sm2_ep_allocation_entry *entries =
+		sm2_mmap_entries(&sm2_av->mmap);
+	struct sm2_cma_data *cma_data =
+		(struct sm2_cma_data *) xfer_entry->user_data;
+
+	pid_t pid = entries[xfer_entry->hdr.sender_gid].pid;
+
+	while (1) {
+		if (write)
+			ret = ofi_process_vm_writev(
+				pid, rx_entry->iov, rx_entry->count,
+				cma_data->iov, cma_data->iov_count, 0);
+		else
+			ret = ofi_process_vm_readv(
+				pid, rx_entry->iov, rx_entry->count,
+				cma_data->iov, cma_data->iov_count, 0);
+		if (ret < 0) {
+			if (errno == 14) {
```

Comment:
I think this should test `errno == EFAULT` rather than 14.

Reply:
Fixed